feat: enable log forwarding through otel collector (#186)
canonical/github-runner-operators: docs/how-to/enable-log-forwarding.md

Reviewer: You'll need to add this file to the `toctree` of in [`docs/how-to/index.rst`](https://github.com/canonical/github-runner-operators/blob/main/docs/how-to/index.rst) -- that should resolve the build errors

# Enable log forwarding

> This action allows workflow authors to opt in to forwarding specific log files from self-hosted GitHub runners to Loki through the OpenTelemetry Collector snap.

Reviewer: Since this is the first sentence of the guide, I think we should be explicit about what action we're referring to:
```suggestion
The `enable-log-forwarding` action allows workflow authors to opt in to forwarding specific log files from self-hosted GitHub runners to Loki through the OpenTelemetry Collector snap.
```

By default, nothing is forwarded. Log forwarding starts only when this action is used in a workflow.

> ## Inputs

Reviewer: Since this is a how-to guide, I prefer if we use instructional headers, so maybe something like:
```suggestion
## Provide inputs
```
or
```suggestion
## Set inputs
```
I'm open to other ideas if you have any!

- `files` (required): newline or comma separated file paths or glob patterns.
- `config-file-name` (optional, default `90-github-runner-log-forwarding.yaml`): generated config file name.
- `otlp-endpoint` (optional): OTLP/HTTP endpoint used to create the exporter when one is not already configured.

> When `otlp-endpoint` is not set, the action falls back to `OTEL_EXPORTER_OTLP_LOGS_ENDPOINT`, then `OTEL_EXPORTER_OTLP_ENDPOINT` from the workflow environment.

Reviewer: No comment on Copilot's comment, but I do have a question about this sentence. What happens if `otlp-endpoint`, `OTEL_EXPORTER_OTLP_LOGS_ENDPOINT` and `OTEL_EXPORTER_OTLP_ENDPOINT` are all unset? Does the action fail?

> When `otlp-endpoint` is not set, the action falls back to `OTEL_EXPORTER_OTLP_LOGS_ENDPOINT`, then `OTEL_EXPORTER_OTLP_ENDPOINT` from the workflow environment.

Author: copilot is right. when the fallback is unset the action will print out the error and should exit (fail)

> ## Usage

Reviewer: Similar comment on this header, can we make it more instructional like:
```suggestion
## Use the action
```
I'm open to ideas if you have any!

```yaml
jobs:
  chrony-testing:

>     runs-on: ubuntu-latest

Author: prerequisites section has now been added

    steps:
      - uses: canonical/github-runner-operators/actions/enable-log-forwarding@main
        with:
          files: |
            /var/log/chrony/*.log
            /var/log/syslog
      - run: ./run-tests.sh

> ```

Reviewer: There's no indication about what to do with this snippet in the guide. Do I need to define a new file? Where does the file go? Is this code snippet complete, or do I need to include additional aspects (e.g., the inputs)?

Author: i have added a sentence before this to instruct the user to add it to a job in their workflow file

Pin to a release tag or commit SHA in production workflows.

> 

Reviewer: At this point in the guide I would like to know: How I can validate that the logs were successfully forwarded? (e.g., does the action output a success message? Does the opentelemetry-collector snap experience an observable change?)

Author: i have added a section to check the logs are forwarded

> ## Loki query hints

Reviewer: Similar comment here about making the section header more instructional, how about:
```suggestion
## Examine Loki queries
```
Let me know if you have any other ideas :) (This might not even be accurate to the content, please let me know!)

The action adds GitHub context as resource attributes on forwarded logs:

- `github.job.id`
- `github.job.name`
- `github.repository`
- `github.runner.name`
- `github.workflow`
- `github.run.attempt`

Example Loki query by workflow run id:

> ```logql

Reviewer: The PR preview of the RTD site is throwing a warning at this line (which failed the build):

```
/home/docs/checkouts/readthedocs.org/user_builds/canonical-github-runner-operators/checkouts/186/docs/how-to/enable-log-forwarding.md:45: WARNING: Pygments lexer name 'logql' is not known
```

So I think you'll need to remove the `logql` part of this line (I dare not add a suggestion, because I don't trust GitHub to correctly apply the suggestion)

{github_job_id="123456789"}
```

## Notes

- This action requires root privileges to write collector config.

> - The `opentelemetry-collector` snap must be installed on the runner.

Reviewer: I suggest we move these notes to the top of the document, as the first section, and rename the section as `## Prerequisites`

# How to enable log forwarding

The `enable-log-forwarding` action allows workflow authors to opt in to forwarding specific log files from self-hosted GitHub runners to Loki through the OpenTelemetry Collector snap.

By default, nothing is forwarded. Log forwarding starts only when this action is used in a workflow.

## Prerequisites

- Use a self-hosted Linux runner.

> - Install the `opentelemetry-collector` snap on the runner.

Author: won't be needed if we install it for the runner in the action steps

- Ensure the workflow can update `/etc/otelcol/config.d` with root privileges.

## Provide inputs

> 

Author: ensure that you do not include any files that may contain sensitive information so that these info are not forwarded

> - Ensure the workflow can update `/etc/otelcol/config.d` with root privileges.
> 
> ## Provide inputs
> 
> - `files` (required): newline or comma separated file paths or glob patterns.

Reviewer: Nitpick, but I think we should provide some instruction about where to provide these inputs, something like:
```suggestion
## Provide inputs

To enable log forwarding, set the following inputs in your workflow file as required by your setup:

- `files` (required): newline or comma separated file paths or glob patterns.
```

- `config-file-name` (optional, default `90-github-runner-log-forwarding.yaml`): generated config file name.
- `otlp-endpoint` (optional): OTLP/gRPC endpoint used to create the exporter when one is not already configured.

When `otlp-endpoint` is not set, the action falls back to `ACTION_OTEL_EXPORTER_OTLP_ENDPOINT` from the workflow environment.

## Use the action

Add this snippet to a job in your workflow file (for example, `.github/workflows/ci.yaml`):

```yaml
jobs:
  chrony-testing:
    runs-on: [self-hosted, linux]
    steps:
      - uses: canonical/github-runner-operators/actions/enable_log_forwarding@main
        with:
          files: |
            /var/log/chrony/*.log
            /var/log/syslog
```

> Pin to a release tag or commit SHA in production workflows.

Reviewer: ```suggestion
Pin the action to a release tag or commit SHA in production workflows.
```
Nitpick, to be explicit about what the user should pin
